src/flint/types/fmpz_mod_poly.pyx

```diff
@@ -0,0 +1,1598 @@
+            >>> f.divmod(g)
+            (21, 106*x^5 + 156*x^4 + 131*x^3 + 43*x^2 + 86*x + 16)
+        """
+        cdef fmpz_t f
+        cdef fmpz_mod_poly Q, R
+
+        other = self.ctx.any_as_fmpz_mod_poly(other)
+        if other is NotImplemented:
+            raise TypeError(f"Cannot interpret {other} as a polynomial")
+
+        if other == 0:
+            raise ZeroDivisionError(f"Cannot compute divmod as {other =}")
+
+        Q = fmpz_mod_poly.__new__(fmpz_mod_poly, None, self.ctx)
+        R = fmpz_mod_poly.__new__(fmpz_mod_poly, None, self.ctx)
+
+        fmpz_init(f)
+        fmpz_mod_poly_divrem_f(
+            f, Q.val, R.val, self.val, (<fmpz_mod_poly>other).val, self.ctx.mod.val
+        )
+        if not fmpz_is_one(f):
+            fmpz_clear(f)
+            raise ValueError(
+                f"Cannot compute divmod of {self} with {other}"
+            )
+
+        return Q, R
+
+    def __divmod__(self, other):
+        return self.divmod(other)
+
+    def __rdivmod__(self, other):
+        other = self.ctx.any_as_fmpz_mod_poly(other)
+        if other is NotImplemented:
+            return other
+        return other.divmod(self)
+
+    def gcd(self, other):
+        """
+        Return the greatest common divisor of self and other.
+
+            >>> R = fmpz_mod_poly_ctx(163)
+            >>> x = R.gen()
+            >>> f = x*(x + 1)
+            >>> f.gcd(x+1)
+            x + 1
+            >>> f.gcd(x*x)
+            x
+
+        """
+        cdef fmpz_mod_poly res
+
+        if not self.ctx.is_prime():
+            raise NotImplementedError("gcd algorithm assumes that the modulus is prime")
+                 
+        other = self.ctx.any_as_fmpz_mod_poly(other)
+        if other is NotImplemented:
+            raise TypeError(f"Cannot interpret {other} as a polynomial")
+
+        res = fmpz_mod_poly.__new__(fmpz_mod_poly, None, self.ctx)
+        fmpz_mod_poly_gcd(
+            res.val, self.val, (<fmpz_mod_poly>other).val, self.ctx.mod.val
+        )
+        return res
+
+    def xgcd(self, other):
+        r"""
+        Computes the extended gcd of self and other: (`G`, `S`, `T`)
+        where `G` is the ``gcd(self, other)`` and `S`, `T` are such that:
+
+        :math:`G = \textrm{self}*S +  \textrm{other}*T`
+
+            >>> R = fmpz_mod_poly_ctx(163)
+            >>> f = R([143, 19, 37, 138, 102, 127, 95])
+            >>> g = R([139, 9, 35, 154, 87, 120, 24])
+            >>> f.xgcd(g)
+            (x^3 + 128*x^2 + 123*x + 91, 17*x^2 + 49*x + 104, 21*x^2 + 5*x + 25)
+
+        """
+        cdef fmpz_mod_poly G, S, T
+        cdef fmpz_t f
+
+        other = self.ctx.any_as_fmpz_mod_poly(other)
+        if other is NotImplemented:
+            raise TypeError(f"Cannot interpret {other} as a polynomial")
+
+        G = fmpz_mod_poly.__new__(fmpz_mod_poly, None, self.ctx)
+        S = fmpz_mod_poly.__new__(fmpz_mod_poly, None, self.ctx)
+        T = fmpz_mod_poly.__new__(fmpz_mod_poly, None, self.ctx)
```

Comment:
I would make a staticmethod on ctx for this:
```python
@staticmethod
cdef fmpz_mod_poly new(fmpz_mod_poly_ctx self):
    return fmpz_mod_poly.__new__(fmpz_mod_poly, None, self)
```
Then it is something like:
```python
cdef fmpz_mod_poly ctx
cdef fmpz_mod_poly G, S, T

ctx = self.ctx
G = ctx.new()
S = ctx.new()
T = ctx.new()
```
I would check how the generated C code looks to see exactly what is happening though.

Reply:
Sorry, not a staticmethod, just an ordinary method. I was going to suggest a staticmethod on `fmpz_mod_poly` like
```python
G = fmpz_mod_poly.new(ctx)
```
but then I realised that it makes more sense as an instance method on the context:
```python
G = ctx.new()
```